source/opt/merge_return_pass.cpp

```diff
@@ -318,7 +315,39 @@ void MergeReturnPass::CreatePhiNodesForInst(BasicBlock* merge_block,
       phi_operands.push_back(pred_id);
     }
 
-    Instruction* new_phi = builder.AddPhi(inst.type_id(), phi_operands);
+    Instruction* new_phi = nullptr;
+    // If the instruction is a pointer and variable pointers are not an option,
+    // then we have to regenerate the instruction instead of creating an OpPhi
+    // instruction.  If not, the Spir-V will be invalid.
+    bool regenerateInstruction =
+        get_def_use_mgr()->GetDef(inst.type_id())->opcode() == SpvOpTypePointer;
+    regenerateInstruction &= !context()->get_feature_mgr()->HasCapability(
+        SpvCapabilityVariablePointers);
```

Comment:
Variable pointers are only allowed for Workgroup and StorageBuffer storage classes. You should probably tighten the restriction.

Reply:
Done.